### PR TITLE
Add details for field.Required

### PR DESCRIPTION
**Pull request by sgallagher**
Some fields are only required if other fields have certain values.
This patch adds some additional error messaging to inform the user
why these fields were necessary.

Fixes issue #7118

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

Replaces https://github.com/openshift/origin/pull/8269

@deads2k PTAL

### CODE REVIEW

File: pkg/cmd/server/api/validation/validation.go [hunk 1]
```diff
@@ -76,15 +76,20 @@ func ValidateHostPort(value string, fldPath *field.Path) field.ErrorList {
...
-			allErrs = append(allErrs, field.Required(fldPath.Child("certFile"), ""))
+			allErrs = append(allErrs, field.Required(fldPath.Child("certFile"), "The certificate file must be provided for this request"))
 		}
 		if len(certInfo.KeyFile) == 0 {
-			allErrs = append(allErrs, field.Required(fldPath.Child("keyFile"), ""))
+			allErrs = append(allErrs, field.Required(fldPath.Child("keyFile"), "The certificate key must be provided for this request"))
 		}
 	}
 
+	if (len(certInfo.CertFile) > 0 && len(certInfo.KeyFile) == 0) || (len(certInfo.CertFile) == 0 && len(certInfo.KeyFile) > 0) {
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/validation.go]: is this `(len(certInfo.CertFile) == 0) != (len(certInfo.KeyFile) == 0)` ?

File: pkg/image/api/validation/validation.go [hunk 2]
```diff
@@ -110,9 +110,6 @@ func ValidateImageStream(stream *api.ImageStream) field.ErrorList {
 func ValidateImageStreamTagReference(tagRef api.TagReference, fldPath *field.Path) field.ErrorList {
 	var errs field.ErrorList
 	if tagRef.From != nil {
-		if len(tagRef.From.Name) == 0 {
```
Comment (deads2k) [inline pkg/image/api/validation/validation.go]: Why was this removed?

Comment (sgallagher) [inline pkg/image/api/validation/validation.go]: See https://github.com/openshift/origin/pull/8269

You asked me to remove it.

File: pkg/cmd/server/api/validation/validation.go (hunk 1 above)
Comment (sgallagher) [inline pkg/cmd/server/api/validation/validation.go]: Damned boolean logic. Yes, this does reduce to that. I'll tweak it.

Comment (liggitt) [inline pkg/cmd/server/api/validation/validation.go]: boolean logic can be tricky, true or false?

File: pkg/image/api/validation/validation.go (hunk 2 above)
Comment (liggitt) [inline pkg/image/api/validation/validation.go]: I think he meant the details string, not the entire check and error

File: pkg/cmd/server/api/validation/validation.go (hunk 1 above)
Comment (sgallagher) [inline pkg/cmd/server/api/validation/validation.go]: False, due to the use of the phrase "can be".

Comment (liggitt) [inline pkg/cmd/server/api/validation/validation.go]: wrong, the answer is true, since true or false always evaluates to true ;)

**Comment by sgallagher** [comment]
@deads2k PTAL

File: pkg/user/api/validation/validation.go [hunk 3]
```diff
@@ -155,10 +155,10 @@ func ValidateIdentity(identity *api.Identity) field.ErrorList {
 		allErrs = append(allErrs, field.Invalid(userPath.Child("name"), identity.User.Name, msg))
 	}
 	if len(identity.User.Name) == 0 && len(identity.User.UID) != 0 {
-		allErrs = append(allErrs, field.Invalid(userPath.Child("uid"), identity.User.UID, "may not be set if user.name is empty"))
+		allErrs = append(allErrs, field.Invalid(userPath.Child("uid"), identity.User.UID, "UID is required when username is provided"))
 	}
 	if len(identity.User.Name) != 0 && len(identity.User.UID) == 0 {
-		allErrs = append(allErrs, field.Required(userPath.Child("uid"), ""))
+		allErrs = append(allErrs, field.Required(userPath.Child("uid"), "username is required when UID is provided"))
```
Comment (liggitt) [inline pkg/user/api/validation/validation.go]: this message doen't make sense... they provided a username

File: pkg/user/api/validation/validation.go [hunk 4]
```diff
@@ -155,10 +155,10 @@ func ValidateIdentity(identity *api.Identity) field.ErrorList {
 		allErrs = append(allErrs, field.Invalid(userPath.Child("name"), identity.User.Name, msg))
 	}
 	if len(identity.User.Name) == 0 && len(identity.User.UID) != 0 {
-		allErrs = append(allErrs, field.Invalid(userPath.Child("uid"), identity.User.UID, "may not be set if user.name is empty"))
+		allErrs = append(allErrs, field.Invalid(userPath.Child("uid"), identity.User.UID, "UID is required when username is provided"))
```
Comment (liggitt) [inline pkg/user/api/validation/validation.go]: this is also phrased oddly, they did provide a UID

Comment (sgallagher) [inline pkg/user/api/validation/validation.go]: Yes, I had those reversed by mistake.

**Comment by sgallagher** [comment]
@deads2k, @liggitt  ping?

File: pkg/cmd/server/api/validation/validation.go (hunk 1 above)
Comment (deads2k) [inline pkg/cmd/server/api/validation/validation.go]: It didn't reduce?

File: pkg/user/api/validation/validation.go [hunk 5]
```diff
@@ -155,10 +155,10 @@ func ValidateIdentity(identity *api.Identity) field.ErrorList {
 		allErrs = append(allErrs, field.Invalid(userPath.Child("name"), identity.User.Name, msg))
 	}
 	if len(identity.User.Name) == 0 && len(identity.User.UID) != 0 {
-		allErrs = append(allErrs, field.Invalid(userPath.Child("uid"), identity.User.UID, "may not be set if user.name is empty"))
+		allErrs = append(allErrs, field.Invalid(userPath.Child("uid"), identity.User.UID, "username is required when UID is provided"))
```
Comment (deads2k) [inline pkg/user/api/validation/validation.go]: something went weird here.  If username is required, why not `field.Required(userPath.Child("username"),....)` ?

**Comment by sgallagher** [comment]
@deads2k Looks like I goofed and pushed an old branch by mistake. This should address your comments from this morning.

File: pkg/user/api/validation/validation.go (hunk 4 above)
Comment (deads2k) [inline pkg/user/api/validation/validation.go]: > Yes, I had those reversed by mistake.

still looks reversed.

File: pkg/user/api/validation/validation.go (hunk 3 above)
Comment (deads2k) [inline pkg/user/api/validation/validation.go]: > this message doesn't make sense... they provided a username

still outstanding.

File: pkg/user/api/validation/validation.go (hunk 4 above)
Comment (sgallagher) [inline pkg/user/api/validation/validation.go]: This is what I get for switching machines partway through a patch...

Should be fixed in the latest push.

File: pkg/user/api/validation/validation.go [hunk 6]
```diff
@@ -155,10 +155,10 @@ func ValidateIdentity(identity *api.Identity) field.ErrorList {
 		allErrs = append(allErrs, field.Invalid(userPath.Child("name"), identity.User.Name, msg))
 	}
 	if len(identity.User.Name) == 0 && len(identity.User.UID) != 0 {
-		allErrs = append(allErrs, field.Invalid(userPath.Child("uid"), identity.User.UID, "may not be set if user.name is empty"))
+		allErrs = append(allErrs, field.Invalid(userPath.Child("uid"), identity.User.UID, "username is required when username is provided"))
```
Comment (deads2k) [inline pkg/user/api/validation/validation.go]: Phrasing this way is good, but you need to make it `field.Required(userpath.Child("username")...)` to match.

File: pkg/user/api/validation/validation.go [hunk 7]
```diff
@@ -155,10 +155,10 @@ func ValidateIdentity(identity *api.Identity) field.ErrorList {
...
 	}
 	if len(identity.User.Name) != 0 && len(identity.User.UID) == 0 {
-		allErrs = append(allErrs, field.Required(userPath.Child("uid"), ""))
+		allErrs = append(allErrs, field.Required(userPath.Child("uid"), "UID is required when UID is provided"))
```
Comment (deads2k) [inline pkg/user/api/validation/validation.go]: match case on uid

File: pkg/user/api/validation/validation.go [hunk 8]
```diff
@@ -155,10 +155,10 @@ func ValidateIdentity(identity *api.Identity) field.ErrorList {
 		allErrs = append(allErrs, field.Invalid(userPath.Child("name"), identity.User.Name, msg))
 	}
 	if len(identity.User.Name) == 0 && len(identity.User.UID) != 0 {
-		allErrs = append(allErrs, field.Invalid(userPath.Child("uid"), identity.User.UID, "may not be set if user.name is empty"))
+		allErrs = append(allErrs, field.Invalid(userPath.Child("username"), "username is required when uid is provided"))
```
Comment (deads2k) [inline pkg/user/api/validation/validation.go]: `field.Required`?

**Comment by pweil-** [comment]
rebased and addressed the remaining feedback in https://github.com/openshift/origin/pull/11034.  Closing this in favor of the new PR.
